Android/src/org/droidplanner/android/maps/providers/baidu_map/BaiduMapOfflineMapPreference.java

```diff
@@ -0,0 +1,357 @@
+        button.setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                clickLocalMapListButton(v);
+            }
+        });
+    }
+
+    /**
+     * switch to city list
+     *
+     * @param view
+     */
+    public void clickCityListButton(View view) {
+        if (mBindDialogView != null) {
+            mLocalMapLayout.setVisibility(View.GONE);
+            mCityListLayout.setVisibility(View.VISIBLE);
+        }
+    }
+
+    /**
+     * switch to download manager list
+     *
+     * @param view
+     */
+    public void clickLocalMapListButton(View view) {
+        if (mBindDialogView != null) {
+            mLocalMapLayout.setVisibility(View.VISIBLE);
+            mCityListLayout.setVisibility(View.GONE);
+        }
+    }
+
+    /**
+     * search offline city
+     *
+     * @param view
+     */
+    public void search(View view) {
+        ArrayList<MKOLSearchRecord> records = mOffline.searchCity(mCityNameView
+                .getText().toString());
+        if (records == null || records.size() != 1) {
+            return;
+        }
+        mCidView.setText(String.valueOf(records.get(0).cityID));
+    }
+
+    /**
+     * Start to download
+     *
+     * @param view
+     */
+    public void start(View view) {
+        int cityid = Integer.parseInt(mCidView.getText().toString());
+        mOffline.start(cityid);
+        clickLocalMapListButton(view);
+        final Context context = getContext();
+        Resources res = context.getResources();
+        String text = String.format(res.getString(R.string.pref_baidu_map_offline_download_start_message), cityid);
+        Toast.makeText(context, text, Toast.LENGTH_SHORT).show();
+        updateView();
+    }
+
+    /**
+     * Pause download
+     *
+     * @param view
+     */
+    public void stop(View view) {
+        int cityid = Integer.parseInt(mCidView.getText().toString());
+        mOffline.pause(cityid);
+        final Context context = getContext();
+        Resources res = context.getResources();
+        String text = String.format(res.getString(R.string.pref_baidu_map_offline_download_stop_message), cityid);
+        Toast.makeText(context, text, Toast.LENGTH_SHORT).show();
+        updateView();
+    }
+
+    /**
+     * Remove offline map
+     *
+     * @param view
+     */
+    public void remove(View view) {
+        int cityid = Integer.parseInt(mCidView.getText().toString());
+        mOffline.remove(cityid);
+        final Context context = getContext();
+        Resources res = context.getResources();
+        String text = String.format(res.getString(R.string.pref_baidu_map_offline_download_remove_message), cityid);
+        Toast.makeText(context, text, Toast.LENGTH_SHORT).show();
+        updateView();
+    }
+
+    /**
+     * Display offline map update status
+     */
+    public void updateView() {
+        mLocalMapList = mOffline.getAllUpdateInfo();
```

Comment:
@droneboost Change the type of `mLocalMapList` to `List<MKOLUpdateElement>` instead of `ArrayList<...>` since that's the interface they all have in common.